opm/simulators/wells/WellState.cpp

```diff
@@ -53,144 +54,71 @@ void WellState::base_init(const std::vector<double>& cellPressures,
-                    break;
-                case InjectorType::MULTI:
-                    // Not currently handled, keep zero init.
-                    break;
-                }
-            } else {
-                // Keep zero init.
-            }
-        } else {
-            assert(well.isProducer());
-            // Note negative rates for producing wells.
-            switch (prod_controls.cmode) {
-            case Well::ProducerCMode::ORAT:
-                assert(pu.phase_used[BlackoilPhases::Liquid]);
-                rates[pu.phase_pos[BlackoilPhases::Liquid]] = -prod_controls.oil_rate;
-                break;
-            case Well::ProducerCMode::WRAT:
-                assert(pu.phase_used[BlackoilPhases::Aqua]);
-                rates[pu.phase_pos[BlackoilPhases::Aqua]] = -prod_controls.water_rate;
-                break;
-            case Well::ProducerCMode::GRAT:
-                assert(pu.phase_used[BlackoilPhases::Vapour]);
-                rates[pu.phase_pos[BlackoilPhases::Vapour]] = -prod_controls.gas_rate;
-                break;
-            default:
-                // Keep zero init.
-                break;
-            }
-        }
-        // 2. Bhp: initialize bhp to be target pressure if
-        //    bhp-controlled well, otherwise set to a
-        //    little above or below (depending on if
-        //    the well is an injector or producer)
-        //    pressure in first perforation cell.
-        if (is_bhp) {
-            ws.bhp = bhp_limit;
-        } else {
-            const double safety_factor = well.isInjector() ? 1.01 : 0.99;
-            ws.bhp = safety_factor * global_pressure;
-        }
-    }
-
-    // 3. Thp: assign thp equal to thp target/limit, if such a limit exists,
-    //    otherwise keep it zero.
-    const bool has_thp = well.isInjector() ? inj_controls.hasControl(Well::InjectorCMode::THP)
-                                           : prod_controls.hasControl(Well::ProducerCMode::THP);
-    const double thp_limit = well.isInjector() ? inj_controls.thp_limit : prod_controls.thp_limit;
-    if (has_thp) {
-        ws.thp = thp_limit;
-    }
+    ws.update_injector_targets(well, summary_state);
 }
 
 
 
+void WellState::initSingleWell(const std::vector<double>& cellPressures,
+                               const Well& well,
+                               const std::vector<PerforationData>& well_perf_data,
+                               const ParallelWellInfo& well_info,
+                               const SummaryState& summary_state)
+{
+    if (well.isInjector() == well.isProducer())
+        throw std::logic_error(fmt::format("Well must be either producer or injector - logic error for well: {}", well.name()));
+
+    double pressure_first_connection;
+    if (well_perf_data.empty())
+        pressure_first_connection = -1;
+    else
+        pressure_first_connection = well_info.broadcastFirstPerforationValue(cellPressures[well_perf_data[0].cell_index]);
```

Comment:
This assumes that for distributed wells if one process sharing the well has no perforations then there are now perforations for this well (empty on all processes sharing the well). As perforations might appear/disappear that does not seem to hold and hence might create deadlocks/bugs.

How about this
```
double pressure_first_connection = -1;
if (!well_perf_data.empty())
  pressure_first_connection =  cellPressures[well_perf_data[0].cell_index];
pressure_first_connection = well_info.broadcastFirstPerforationValue(pressure_first_connection);
```

Reply:
Thanks - I'll update!

Reply:
NB: I have marked this PR as draft - and will update the refactor-only-PR #3659